src/domain/mongoose_domain_api.erl

```diff
@@ -0,0 +1,130 @@
+%% Main module other parts of MongooseIM should use to access the domain
+%% management.
+-module(mongoose_domain_api).
+
+-export([init/1,
+         insert_domain/2,
+         remove_domain/1,
+         disable_domain/1,
+         enable_domain/1,
+         get_host_type/1,
+         get_all_locked/0,
+         get_domains_by_host_type/1]).
+
+-type domain() :: jid:lserver().
+-type host_type() :: jid:lserver().
+-type pair() :: {domain(), host_type()}.
+
+%% Init - on init all the “enabled” domain names from the persistent storage
+%% must be added to the core MIM component described above.
+%% Domains should be nameprepped using `jid:nameprep'
+-spec init([pair()]) -> ok | {error, term()}.
+init(Pairs) ->
+    mongoose_domain_core:start(Pairs).
+
+%% Add domain name (w/ host type) - This function must be idempotent.
+%% Added domain is always “enabled” by default it must be added in the core
+%% MIM component described in the previous section.
+%% If it’s successfully enabled than Information about the domain name
+%% must be added into persistent storage and distributed across all the nodes
+%% in the cluster.
+%% Domain and HostType should be nameprepped using `jid:nameprep'
+-spec insert_domain(domain(), host_type()) ->
+    ok  | {error, duplicate} | {error, {db_error, term()}}
+    | {error, service_disabled}.
+insert_domain(Domain, HostType) ->
+    case mongoose_domain_core:is_locked(Domain) of
+        true ->
+            {error, duplicate};
```

Comment:
all other places return `{error, locked}` error in this case, let's be consistent and return the same here.